### PR TITLE
Update `io/pdf.py` to new pypdfium2 API

### PR DESCRIPTION
**Pull request by mara004**
This patch would make doctr compatible with pypdfium2 >= 2.0.0, which will be released soon.
Now that pypdfium2 supports extracting text and locating images, it would also be possible to add back some features that were removed with https://github.com/mindee/doctr/pull/829.

### CODE REVIEW

**Comment by mara004** [comment]
CI currently fails because of the >= 2.0.0 requirement, considering that this version is not actually released yet.

**Comment by felixdittrich92** [comment]
@mara004 Thanks a lot for this update :+1:

**Comment by felixdittrich92** [comment]
@mara004 thanks only a few minor things left :+1:

File: doctr/io/pdf.py [hunk 1]
```diff
@@ -35,8 +34,9 @@ def read_pdf(file: AbstractFile, scale: float = 2, **kwargs: Any) -> List[np.nda
     if not isinstance(file, (str, bytes)):
         raise TypeError("unsupported object type for argument 'file'")
 
-    if isinstance(file, str) and not os.path.isfile(file):
-        raise FileNotFoundError(f"unable to access {file}")
-
     # Rasterise pages to PIL images with pypdfium2 and convert to numpy ndarrays
-    return [np.asarray(img) for img, _ in pdfium.render_pdf_topil(file, scale=scale, **kwargs)]
+    pdf = pdfium.PdfDocument(file, password=password)
```
[the remark this answers is not in the document]
Comment (mara004) [reply doctr/io/pdf.py:38]: Sorry, that's not available yet. PdfDocument currently doesn't implement `__enter__()` and `__exit__()`.

**Review by felixdittrich92 — CHANGES_REQUESTED** [review]

File: doctr/io/pdf.py [hunk 2]
```diff
@@ -15,7 +14,12 @@
 __all__ = ['read_pdf']
 
 
-def read_pdf(file: AbstractFile, scale: float = 2, **kwargs: Any) -> List[np.ndarray]:
+def read_pdf(
+        file: AbstractFile,
+        scale: float = 2,
+        password: Optional[str] = None,
+        **kwargs: Any,
+    ) -> List[np.ndarray]:
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/io/pdf.py:22]:
```
def read_pdf(
    file: AbstractFile, 
    scale: float = 2, 
    password: Optional[str] = None, 
    **kwargs: Any,
) -> List[np.ndarray]:
```
you can check the style also with `flake8 .` in the project directory :+1:

**Review by felixdittrich92 — APPROVED** [review]
Thanks a lot :)

**Comment by mara004** [comment]
Thank you for reviewing/merging!

**Review by frgfm — COMMENTED** [review]
Thanks a lot @mara004 :)

Just a few comments & questions out of curiosity!

File: doctr/io/pdf.py [hunk 3]
```diff
@@ -15,7 +14,12 @@
 __all__ = ['read_pdf']
 
 
-def read_pdf(file: AbstractFile, scale: float = 2, **kwargs: Any) -> List[np.ndarray]:
+def read_pdf(
+    file: AbstractFile,
+    scale: float = 2,
+    password: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (mara004) [reply doctr/io/pdf.py:20]: I agree the term "Optional" is confusing for required parameters, but as `password` is actually optional, I wouldn't worry about that here. What about keeping it as-is for now and switching to `str | None` once you require Python >=3.10 ?

File: doctr/io/pdf.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mara004) [reply doctr/io/pdf.py:44]: As you can see in https://github.com/mindee/doctr/pull/953, `PdfDocument` now provides the context manager API, so you can use any instance of this class in a `with` block, meaning that `.close()` will be called automatically once the context manager exits. Does this answer your question?

**Pull request merged**